app/uk/gov/hmrc/cgtpropertydisposalsfrontend/controllers/HomeController.scala

```diff
@@ -18,40 +18,59 @@ package uk.gov.hmrc.cgtpropertydisposalsfrontend.controllers
 
 import com.google.inject.{Inject, Singleton}
 import play.api.mvc.{Action, AnyContent, MessagesControllerComponents, Result}
-import uk.gov.hmrc.cgtpropertydisposalsfrontend.config.ViewConfig
+import uk.gov.hmrc.cgtpropertydisposalsfrontend.config.{ErrorHandler, ViewConfig}
 import uk.gov.hmrc.cgtpropertydisposalsfrontend.controllers.actions.{AuthenticatedAction, RequestWithSessionData, SessionDataAction, WithAuthAndSessionDataAction}
 import uk.gov.hmrc.cgtpropertydisposalsfrontend.models.JourneyStatus.Subscribed
-import uk.gov.hmrc.cgtpropertydisposalsfrontend.util.{Logging, toFuture}
+import uk.gov.hmrc.cgtpropertydisposalsfrontend.models.{Error, SessionData}
+import uk.gov.hmrc.cgtpropertydisposalsfrontend.repos.SessionStore
+import uk.gov.hmrc.cgtpropertydisposalsfrontend.util.Logging
+import uk.gov.hmrc.cgtpropertydisposalsfrontend.util.Logging._
 import uk.gov.hmrc.cgtpropertydisposalsfrontend.views
 import uk.gov.hmrc.play.bootstrap.controller.FrontendController
 
-import scala.concurrent.Future
+import scala.concurrent.{ExecutionContext, Future}
 
 @Singleton
 class HomeController @Inject()(
   val authenticatedAction: AuthenticatedAction,
   val sessionDataAction: SessionDataAction,
+  errorHandler: ErrorHandler,
+  sessionStore: SessionStore,
   cc: MessagesControllerComponents,
   manageYourDetailsPage: views.html.account.manage_your_details
-)(implicit viewConfig: ViewConfig)
+)(implicit viewConfig: ViewConfig, ec: ExecutionContext)
     extends FrontendController(cc)
     with WithAuthAndSessionDataAction
     with SessionUpdates
     with Logging {
 
   def homepage(): Action[AnyContent] = authenticatedActionWithSessionData.async {
     implicit request: RequestWithSessionData[AnyContent] =>
-      withSubscribedUser(request) { subscribed =>
-        Ok(manageYourDetailsPage(subscribed.subscribedDetails))
+      withSubscribedUser(request) { (sessionData, subscribed) =>
+        val updateSessionResult =
+          sessionData.subscriptionDetailUpdated.fold[Future[Either[Error, Unit]]](
```

Comment:
I'm not a `cats` expert but I believe `cats-mtl` solves these crappy type inference problems - type inference is one of the major problems with cats monad transfromers

Reply:
I haven't come across the `mtl` library, but it's worth looking into if it'll make things easier. I'll have a look to see what it could give us

Reply:
I'm not sure it would necessarily help in this part of the code. We're not dealing with monad transformers here - the type inference problem arises from the `fold` on the `Option`. The fold tries to derive the target type from the return type of the first `isEmpty` block. Without the type annotation it dervies `Future[Right[Nothing,Unit]]` from `Future.successful(Right(()))` and the expects that return type in the second `ifDefined` block

Reply:
Yes, you're right in this case. I auto-jumped to cats as I was hitting that problem earlier. But this is even worse, what does that say about the Scala type system? :zipper_mouth_face:

Reply:
maybe we should use our own fold on Option with better type inference :laughing:  :
```scala
  def fold[A,B](x: Option[A])(ifEmpty: => B, ifDefined: A => B): B = x.fold(ifEmpty)(ifDefined)
```